homeassistant/components/thethingsnetwork/config_flow.py

```diff
@@ -0,0 +1,117 @@
+"""The Things Network's integration config flow."""
+
+from collections.abc import Mapping
+import logging
+import traceback
+from typing import Any
+
+from ttn_client import TTNAuthError, TTNClient
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry, ConfigFlow, ConfigFlowResult
+from homeassistant.helpers.selector import (
+    TextSelector,
+    TextSelectorConfig,
+    TextSelectorType,
+)
+
+from .const import CONF_API_KEY, CONF_APP_ID, CONF_HOSTNAME, DOMAIN, TTN_API_HOSTNAME
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class TTNFlowHandler(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize flow."""
+        self._reauth_entry: ConfigEntry | None = None
+
+    async def async_step_user(
+        self, user_input: Mapping[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """User initiated config flow."""
+
+        errors = {}
+        if user_input is not None:
+            connection_error = None
+            try:
+                client = TTNClient(
+                    user_input[CONF_HOSTNAME],
+                    user_input[CONF_APP_ID],
+                    user_input[CONF_API_KEY],
+                    0,
+                )
+                await client.fetch_data()
+            except TTNAuthError:
+                # Raising ConfigEntryAuthFailed will cancel future updates
+                # and start a config flow with SOURCE_REAUTH (async_step_reauth)
+                _LOGGER.error("TTNAuthError")
+                connection_error = "invalid_auth"
+            except Exception as err:  # pylint: disable=broad-exception-caught
+                _LOGGER.error(err)
+                _LOGGER.error(traceback.format_exc())
+                connection_error = "unknown"
+
+            if connection_error:
+                errors["base"] = connection_error
+            else:
+                # Create entry
+                if self._reauth_entry:
+                    return self.async_update_reload_and_abort(
+                        self._reauth_entry,
+                        data=user_input,
+                        reason="reauth_successful",
+                    )
+                if not self.unique_id:
+                    await self.async_set_unique_id(user_input[CONF_APP_ID])
```

Comment:
App id can be changed like you said, so this is not a good app id. Can we fetch some kind of user id from the service?